runtime/shared_heap.c

```diff
@@ -1319,64 +1404,747 @@ void caml_compact_heap(caml_domain_state* domain_state,
     }
 
-    pool_freelist.free = NULL;
+    heap->unswept_full_pools[sz_class] = new_full_pools;
+    heap->unswept_avail_pools[sz_class] = new_avail_pools;
+    if( evac_pool_last != NULL ) {
+      CAMLassert(evac_pools);
+      evac_pool_last->next = domain_evac_pools;
+      domain_evac_pools = evac_pools;
+    }
+  }
 
-    caml_plat_unlock(&pool_freelist.lock);
+  CAML_EV_END(EV_COMPACT_EVACUATE);
+  caml_global_barrier(participating_count);
+
+  compact_debug_check_heap(heap);
+
+  caml_global_barrier(participating_count);
+  compact_fix(participants[0] == Caml_state);
+  caml_global_barrier(participating_count);
+
+  /* Third step: move all evacuated pools to the pool freelist */
+
+  CAML_EV_BEGIN(EV_COMPACT_RELEASE);
+  pool* cur_pool = domain_evac_pools;
+  while (cur_pool) {
+    pool* next_pool = cur_pool->next;
+
+    #ifdef DEBUG
+    for (header_t *p = POOL_FIRST_BLOCK(cur_pool, cur_pool->sz);
+        p < POOL_END(cur_pool); p++) {
+      *p = Debug_free_major;
+    }
+    #endif
+
+    pool_release(heap, cur_pool, cur_pool->sz);
+    cur_pool = next_pool;
   }
 
+  CAML_EV_END(EV_COMPACT_RELEASE);
+}
+
+static int should_run_phase_two = 0;
+
+/* New algorithm main driver. Run in parallel for all domains.
+
+   We support two algorithms; "old" and "new". Default is "new".
+
+   This compaction algorithm operates in two phases. The first phase
+   we sort pools in each class size for each domain, prefering larger
+   chunks of pools. We then compact each size class in parallel across
+   all domains, creating a list of free pools.
+
+   The second phase we globally sort all pools and move those so we end
+   up with whole chunks of free pools that can be freed.
+
+   Each phase implements a compaction algorithm that is similar to
+   Edward's Two-Finger algorithm from the original 1974 LISP book (The
+   Programming Language LISP). At a high level the algorithm works as a series
+   of parallel (using all running domains) steps separated by global barriers:
+
+   1. For each size class
+    a. Compute the number of live blocks in partially filled pools
+    b. Keep enough pools to fully contain the number of live blocks and
+       set the rest to be evacuated
+    c. For each live block in each pool in the evacuation list,
+       allocate and copy into a non-evacuating pool.
+  2. Proceed through the roots and the heap, updating pointers to evacuated
+     blocks to point to the new location of the block. Update finalisers and
+     ephemerons too.
+  3. Go through pools evacuated and release them. Finally free all but
+      one pool in the freelist.
+  4. One domain needs to release the pools in the freelist back to the OS.
+*/
+
+static void compact_new_algorithm(caml_domain_state* domain_state,
+                                  int participating_count,
+                                  caml_domain_state** participants)
+{
+  struct caml_heap_state* heap = Caml_state->shared_heap;
+
+  CAML_EV_BEGIN(EV_COMPACT_EVACUATE);
```

Comment:
The handling of `EV_COMPACT_EVACUATE` is wrong in this function - it's begun twice and ended zero or one times.

(I'm surprised test_compact.ml doesn't complain about this - am I misreading something?)

Reply:
Annoyingly, `compact_run_phase` changes the runtime events phase (as documented in [this comment](https://github.com/ocaml-flambda/flambda-backend/pull/3179/files#diff-b468e47c34ab18ea7b2cd8322cde20a2fa4d3e647422ee725f0fe7d84f768a32R1835-R1836). This was frustrating when I wrote it. I'll add a comment here at the call sites.